pkg/networkserver/downlink.go

```diff
@@ -931,25 +965,32 @@ func (ns *NetworkServer) processDownlinkTask(ctx context.Context) error {
 
 				logger = logger.WithField("device_class", dev.MACState.DeviceClass)
 
-				if dev.MACState.DeviceClass == ttnpb.CLASS_A && !dev.MACState.RxWindowsAvailable {
-					logger.Warn("Rx windows not available, skip downlink slot")
-					return dev, nil, nil
-				}
-
-				if dev.MACState.DeviceClass == ttnpb.CLASS_B && !dev.MACState.RxWindowsAvailable {
-					// TODO: Support Class B (https://github.com/TheThingsNetwork/lorawan-stack/issues/19).
-					logger.Warn("Rx windows not available, skip downlink slot")
-					return dev, nil, nil
+				if !dev.MACState.RxWindowsAvailable {
+					logger.Debug("Rx windows not available, skip class A downlink slot")
+					if dev.MACState.DeviceClass == ttnpb.CLASS_A {
+						return dev, nil, nil
+					}
 				}
 
-				if len(dev.RecentUplinks) == 0 && (dev.MACState.DeviceClass == ttnpb.CLASS_A || dev.MACState.DeviceClass == ttnpb.CLASS_B) {
-					logger.Warn("No recent uplinks found, skip downlink slot")
-					return dev, nil, nil
+				var maxUpLength uint16 = math.MaxUint16
+				if dev.MACState.LoRaWANVersion == ttnpb.MAC_V1_1 {
+					maxUpLength = maximumUplinkLength(fp, phy, dev.RecentUplinks...)
 				}
 
-				maxUpLength := maximumUplinkLength(fp, phy, dev.RecentUplinks...)
+				var sets []string
+			outer:
+				for dev.MACState.RxWindowsAvailable {
```

Comment:
This is quite unconventional use of `for`, I use it here to get a way to go to exit early from the block.
Other options:
1. very deeply nested `if`
2. a very specialized named function with several pointer arguments/return values
3. anonymous function executed right on spot (`func(){ /* class A downlink handling */ }()`), which has a negative of confusing `return` statements
4. `goto` - works out-of-the-box, but AFAIK @johanstokking is anti-`goto`

2. 3. suffer from issue of inability to `return` directly from the block(stop the transaction), so additional error handling should be added outside the function

IMO this is a perfect use-case for `goto`

Reply:
I think option 2 is the way to go. This method is already more than 500 lines long

I'm not strictly against this use of `for`, but I have increasing concerns with the maintainability of the NS code base

Reply:
What are the concerns exactly?

Reply:
Mostly the lengthy funcs. I think we do a good job given all the conditions, it's not meant offensive, but code tricks like this make it hard to read, let alone maintain, for somebody else. If a new person becomes code owner of NS at some time in the future, they have to figure out why there's a `for { }` here that's not meant as a loop, and make sure it never reaches an infinite loop.